### PR TITLE
Implement caching layer for block information

### LINKED ISSUE #914: Implement caching layer for block information
Add Block table to the database and make sure it gets updated at every block. The table should contain:

- Block height (primary key)
- Transaction count
- Timestamp
- Proposer (foreign key to validator table)
- BlockHash

### PR DESCRIPTION
**Pull request by prin-r**
Fix #914

### CODE REVIEW

File: chain/db/table.go [hunk 1]
```diff
@@ -46,3 +46,10 @@ type DataSourceRevision struct {
 	BlockHeight    int64  `gorm:"not null"`
 	TxHash         []byte `gorm:"not null"`
 }
+
+type Block struct {
+	Height    int64  `gorm:"primary_key;auto_increment:false"`
+	Timestamp int64  `gorm:"not null"`
+	Proposer  string `gorm:"not null"`
+	BlockHash string `gorm:"unique;not null"`
```
Comment (sorawit) [inline chain/db/table.go]: I don't think we need to say it's unique?

File: chain/db/block.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+func (b *BandDB) AddBlock(
+	height int64,
+	timestamp time.Time,
+	proposer string,
+	blockHash common.HexBytes,
+) error {
+	return b.tx.Create(&Block{
+		Height:    height,
+		Timestamp: timestamp.Unix(),
```
Comment (sorawit) [inline chain/db/block.go]: Let's do more precise timestamp here.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go (hunk 1 above)
Comment (taobun) [reply chain/db/table.go]: We can store block hash as []byte

File: chain/db/block.go (hunk 2 above)
Comment (taobun) [reply chain/db/block.go]: What does it mean?

File: chain/app/db_app.go [hunk 3]
```diff
@@ -157,6 +158,13 @@ func (app *dbBandApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseB
 		panic(err)
 	}
 
+	app.dbBand.AddBlock(
+		req.Header.GetHeight(),
+		time.Now(),
+		fmt.Sprintf("%X", req.Header.GetProposerAddress()),
```
Comment (taobun) [inline chain/app/db_app.go]: Wow please change my code in validator.go too

File: chain/db/block.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"time"
+
+	"github.com/tendermint/tendermint/libs/common"
+)
+
+func (b *BandDB) AddBlock(
+	height int64,
+	timestamp time.Time,
+	proposer string,
```
Comment (taobun) [inline chain/db/block.go]: Receive proposer as cmn.HexBytes and cast to string in function instead.

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 5]
```diff
@@ -157,6 +157,13 @@ func (app *dbBandApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseB
 		panic(err)
 	}
 
+	app.dbBand.AddBlock(
+		req.Header.GetHeight(),
+		time.Now(),
```
Comment (taobun) [inline chain/app/db_app.go]: Use block time from context

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 6]
```diff
@@ -157,6 +157,13 @@ func (app *dbBandApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseB
 		panic(err)
 	}
 
+	app.dbBand.AddBlock(
+		req.Header.GetHeight(),
+		app.DeliverContext.BlockHeader().Time,
```
Comment (taobun) [inline chain/app/db_app.go]: `app.DeliverContext.BlockTime()`

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**